frontend/userpatch.lua

```diff
@@ -0,0 +1,121 @@
+--]]--
+
+local isAndroid, android = pcall(require, "android")
+
+local userpatch = {
+    -- priorities for user patches,
+    early_once = "0",  -- to be started early on startup (once after an update)
+    early = "1",       -- to be started early on startup (always, but after an `early_once`)
+    late = "2",        -- to be started after UIManager is ready (always)
+                       -- 3-7 are reserved for later use
+    before_exit = "8", -- to be started a bit before exit before settings are saved (always)
+    on_exit = "9",     -- to be started right before exit (always)
+
+    -- the patch function itself
+    applyPatches = function(priority) end, -- to be overwritten, if the device allows it.
+}
+
+if isAndroid and android.prop.flavor == "fdroid" then
+    return userpatch -- allows to use applyPatches as a no-op on F-Droid flavor
+end
+
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local DataStorage = require("datastorage")
+
+-- the directory KOReader is installed in (and runs from)
+local package_dir = lfs.currentdir()
+-- the directory where KOReader stores user data
+local data_dir = DataStorage:getDataDir()
+
+--- Run lua patches
+-- Execution order order is alphanum-sort for humans version 4: `1-patch.lua` is executed before `10-patch.lua`
+-- (see http://notebook.kulchenko.com/algorithms/alphanumeric-natural-sorting-for-humans-in-lua)
+-- string directory ... to scan through (flat no recursion)
+-- string priority ... only files starting with `priority` followed by digits and a '-' will be processed.
+-- return true if a patch was executed
+local function runUserPatchTasks(dir, priority)
+    if lfs.attributes(dir, "mode") ~= "directory" then
+        return
+    end
+
+    local patches = {}
+    for entry in lfs.dir(dir) do
+        local mode = lfs.attributes(dir .. "/" .. entry, "mode")
+        if entry and mode == "file" and entry:match("^" .. priority .. "%d*%-") then
+            table.insert(patches, entry)
+        end
+    end
+
+    if #patches == 0 then
+        return -- nothing to do
+    end
+
+    local function addLeadingZeroes(d)
+        local dec, n = string.match(d, "(%.?)0*(.+)")
+        return #dec > 0 and ("%.12f"):format(d) or ("%s%03d%s"):format(dec, #n, n)
+    end
+    local function sorting(a, b)
+        return tostring(a):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#b)
+            < tostring(b):gsub("%.?%d+", addLeadingZeroes)..("%3d"):format(#a)
+    end
+
+    table.sort(patches, sorting)
+
+    for i, entry in ipairs(patches) do
+        local fullpath = dir .. "/" .. entry
+        if lfs.attributes(fullpath, "mode") == "file" then
+            if fullpath:match("%.lua$") then -- execute patch-files first
+                logger.info("User patch apply:", fullpath)
+                local ok, err = pcall(dofile, fullpath)
+                if not ok then
+                    logger.warn("User patch error applying patch:", fullpath, err)
```

Comment:
```suggestion
logger.warn("Patching failed:", err)
```

(Reasoning for omitting the filename is that we've just logged it right above).